### ISSUE TITLE
Forge reports wrong gas usage even with --isolate

### ISSUE THREAD

**Issue by 0xpolarzero** [foundry-rs/foundry#7276]
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (551bcb5c 2024-02-28T08:24:42.597197000Z)

### What command(s) is the bug in?

forge test --isolate

### Operating System

macOS (Apple Silicon)

### Describe the bug

This bug occurs on [commit 551bcb5 (isolated execution)](https://github.com/foundry-rs/foundry/commit/551bcb5c84d672670e1e3fd897089e606d808774), the latest release on 2024-02-29.

## Description

On simple patterns, the new isolated execution mode works fine, from what I experienced.

However, when testing against a more complex contract, i.e. [GasliteDrop1155](https://github.com/0xpolarzero/foundry-gas-report-minimal-repro/blob/main/src/GasliteDrop1155.sol), a contract for airdropping ERC1155 tokens, I noticed some huge discrepancies between Forge report and the reality/Hardhat report.

Basically, an overestimation of ~twice the gas it should use.

## PoC

Here is [a small reproducible example](https://github.com/0xpolarzero/foundry-gas-report-minimal-repro/tree/main)—I tried to keep it minimal.

In this example, the opposite happens actually. Meaning, Forge underestimates gas consumption, when on my larger test it overestimated it. But that should be enough to surface the issue.

The example is a minimal ERC1155 implementation, with a single token sent to a single recipient. After cloning the repo and installing dependencies (npm-style), you can check the console after running both:
- `pnpm hadhat test`;
- `forge test -vvvv --isolate`.

**Comment by Evalir** [comment]
@0xpolarzero thanks for this! is there _any_ chance you could also provide a repro for the overestimation case? We're not expecting isolate to still be 100% perfect, so having these is great for ironing out any kinks.

**Comment by klkvr** [comment]
I have copy-pasted code of the `GasliteDrop1155Test` to a script, and ran it on sepolia.

Resulted gas used by the [transaction](https://sepolia.etherscan.io/tx/0x1a142eb649505f19c9b6371930444e08824378ba9d8ae488770808e8650cd171) exactly matches forge report (57327), so in this case it seems to be an incorrectly reported usage by hardhat

**Comment by 0xpolarzero** [comment]
So I finally have something more accurate to show.

First, the values reported for [`GasliteDrop1155`](https://github.com/0xpolarzero/foundry-gas-report-minimal-repro/blob/main/src/GasliteDrop1155.sol) are perfectly accurate. I was confused by many things, among them, using `--via-ir` because I was too lazy about fixing some arrays copied from memory to storage.

However, I added tests with [Thirdweb's `AirdropERC1155`](https://github.com/0xpolarzero/foundry-gas-report-minimal-repro/blob/main/src/thirdweb/AirdropERC1155.sol), which was initially how I noticed this issue.

While I was wrong about my initial statement, there appears to be an issue indeed with how the Thirdweb contract's usage is measured.

---

Basically, this contract is intended to be used through a proxy, which might be the reason, or at least an indicator. Anyway, I'm not really insighful enough to help, so I tried to provide a convenient way to reproduce it [in the original reproduction repo](https://github.com/0xpolarzero/foundry-gas-report-minimal-repro/tree/main).

Just follow the initial [How to run](https://github.com/0xpolarzero/foundry-gas-report-minimal-repro/tree/main?tab=readme-ov-file#how-to-run) and then go straight to [`--gas-report` won't report accurate values](https://github.com/0xpolarzero/foundry-gas-report-minimal-repro/tree/main?tab=readme-ov-file#--gas-report-wont-report-accurate-values).

Apologies for the organization of the repo and how convoluted the tests might be; I hope it will be clear enough to understand.

---

To document here as well, here is the [gas report](https://github.com/0xpolarzero/foundry-gas-report-minimal-repro/blob/main/gas-report.txt) with only relevant measurements as well as Sepolia transactions:

| src/GasliteDrop1155.sol:GasliteDrop1155 contract |                 |          |          |          |         |
|--------------------------------------------------|-----------------|----------|----------|----------|---------|
| Function Name                                    | min             | avg      | median   | max      | # calls |
| airdropERC1155                                   | 14934491        | 14934491 | 14934491 | 14934491 | 1       |

[Sepolia tx: 14,934,491](https://sepolia.etherscan.io/tx/0x319180ecb6421e686df92e3652c58852e3f28c19261c272676f4a4f8b5618d68)

| src/thirdweb/AirdropERC1155.sol:AirdropERC1155 contract |                 |          |          |          |         |
|---------------------------------------------------------|-----------------|----------|----------|----------|---------|
| Function Name                                           | min             | avg      | median   | max      | # calls |
| airdropERC1155                                          | 14802928        | 14802928 | 14802928 | 14802928 | 1       |

[Sepolia tx: 15,205,153](https://sepolia.etherscan.io/tx/0x4f0f67b7206467d1ee47681552217eaa0df0b03df77adbcbca9db6098f217e8e)

**Comment by 0xpolarzero** [comment]
@klkvr That's interesting, good to keep in mind.

Thanks a lot for taking the time to explain! I'll close the issue then.
